src/Xamarin.Android.Build.Tasks/Tasks/GenerateJavaStubs.cs

```diff
@@ -332,17 +359,32 @@ void Run (DirectoryAssemblyResolver res)
 			string applicationTemplateFile = "ApplicationRegistration.java";
 			SaveResource (applicationTemplateFile, applicationTemplateFile, real_app_dir,
 				template => template.Replace ("// REGISTER_APPLICATION_AND_INSTRUMENTATION_CLASSES_HERE", regCallsWriter.ToString ()));
+
+#if ENABLE_MARSHAL_METHODS
+			void StoreMarshalAssemblyPath (string name, ITaskItem asm)
+			{
+				if (Debug) {
+					return;
+				}
+
+				if (!marshalMethodsAssemblyPaths.TryGetValue (name, out HashSet<string> assemblyPaths)) {
+					assemblyPaths = new HashSet<string> ();
+					marshalMethodsAssemblyPaths.Add (name, assemblyPaths);
+				}
+
+				if (!assemblyPaths.Contains (asm.ItemSpec)) {
```

Comment:
You don't need `.Contains()` + `.Add()` for a `HashSet<string>`, and the "double-check" is likely *slower* as `.Add()` checks to see if the value is already present *anyway*.